modules/AlphaWalletFoundation/AlphaWalletFoundation/KeyManagement/DefaultsWalletAddressesStore.swift

```diff
@@ -92,42 +92,50 @@ public struct DefaultsWalletAddressesStore: WalletAddressesStore {
         }
     }
 
-    private var didAddWalletSubject: PassthroughSubject<AlphaWallet.Address, Never> = .init()
+    private var didAddWalletSubject: PassthroughSubject<Wallet, Never> = .init()
     private var didRemoveWalletSubject: PassthroughSubject<Wallet, Never> = .init()
 
-    public var didAddWalletPublisher: AnyPublisher<AlphaWallet.Address, Never> {
+    public var didAddWalletPublisher: AnyPublisher<Wallet, Never> {
         didAddWalletSubject.eraseToAnyPublisher()
     }
 
     public var didRemoveWalletPublisher: AnyPublisher<Wallet, Never> {
         didRemoveWalletSubject.eraseToAnyPublisher()
     }
 
-    mutating public func addToListOfWatchEthereumAddresses(_ address: AlphaWallet.Address) {
-        watchAddresses = [watchAddresses, [address.eip55String]].flatMap { $0 }
+    mutating public func add(wallet: Wallet) {
+        switch wallet.origin {
+        case .hd:
+            addToListOfEthereumAddressesWithSeed(wallet.address)
+        case .privateKey:
+            addToListOfEthereumAddressesWithPrivateKeys(wallet.address)
+        case .watch:
+            addToListOfWatchEthereumAddresses(wallet.address)
+        }
+        didAddWalletSubject.send(wallet)
+    }
 
-        didAddWalletSubject.send(address)
+    mutating private func addToListOfWatchEthereumAddresses(_ address: AlphaWallet.Address) {
+        watchAddresses = [watchAddresses, [address.eip55String]].flatMap { $0 }
     }
 
-    mutating public func addToListOfEthereumAddressesWithPrivateKeys(_ address: AlphaWallet.Address) {
+    mutating private func addToListOfEthereumAddressesWithPrivateKeys(_ address: AlphaWallet.Address) {
         let updatedOwnedAddresses = Array(Set(ethereumAddressesWithPrivateKeys + [address.eip55String]))
         ethereumAddressesWithPrivateKeys = updatedOwnedAddresses
-
-        didAddWalletSubject.send(address)
     }
 
-    mutating public func addToListOfEthereumAddressesWithSeed(_ address: AlphaWallet.Address) {
+    mutating private func addToListOfEthereumAddressesWithSeed(_ address: AlphaWallet.Address) {
         let updated = Array(Set(ethereumAddressesWithSeed + [address.eip55String]))
         ethereumAddressesWithSeed = updated
-
-        didAddWalletSubject.send(address)
     }
 
     mutating public func addToListOfEthereumAddressesProtectedByUserPresence(_ address: AlphaWallet.Address) {
         let updated = Array(Set(ethereumAddressesProtectedByUserPresence + [address.eip55String]))
         ethereumAddressesProtectedByUserPresence = updated
 
-        didAddWalletSubject.send(address)
+        //NOTE: not sure if its good, but it shouldn't cause any issues
```

Comment:
Seems ok. Should remove the comment though. It makes more sense now while reviewing the PR, but might be confusing outside of the PR